File: samples/openapi3/client/petstore/spring-cloud-3-with-optional/src/main/java/org/openapitools/configuration/ClientConfiguration.java
package org.openapitools.configuration;

import org.springframework.context.annotation.Bean;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.AnonymousAuthenticationToken;
import org.springframework.security.oauth2.client.AuthorizedClientServiceOAuth2AuthorizedClientManager;
import org.springframework.security.oauth2.client.OAuth2AuthorizeRequest;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientManager;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.security.oauth2.core.OAuth2AccessToken;
import org.springframework.security.core.authority.AuthorityUtils;
import org.springframework.http.HttpHeaders;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;

import feign.RequestInterceptor;
import feign.RequestTemplate;

import org.springframework.context.annotation.Configuration;


@Configuration
public class ClientConfiguration {

  private static final String CLIENT_PRINCIPAL_IMPLICIT = "oauth2FeignClient";

  @Bean
  @ConditionalOnProperty( prefix = "spring.security.oauth2.client.registration.petstoreAuthImplicit", name = "enabled", havingValue = "true" )
  public OAuth2RequestInterceptor implicitOAuth2RequestInterceptor(final OAuth2AuthorizedClientManager implicitAuthorizedClientManager ) {
     return new OAuth2RequestInterceptor(OAuth2AuthorizeRequest.withClientRegistrationId("petstoreAuthImplicit")
            .principal( new AnonymousAuthenticationToken( CLIENT_PRINCIPAL_IMPLICIT, CLIENT_PRINCIPAL_IMPLICIT, AuthorityUtils.createAuthorityList( "ROLE_ANONYMOUS" ) ) )
            .build(), implicitAuthorizedClientManager );
  }

  @Bean
  @ConditionalOnProperty( prefix = "spring.security.oauth2.client.registration.petstoreAuthImplicit", name = "enabled", havingValue = "true" )
  public OAuth2AuthorizedClientManager implicitAuthorizedClientManager(ClientRegistrationRepository clientRegistrationRepository,
        OAuth2AuthorizedClientService authorizedClientService ) {
    return new AuthorizedClientServiceOAuth2AuthorizedClientManager( clientRegistrationRepository, authorizedClientService );
  }
  @Value("${openapipetstore.security.apiKey.key:}")
  private String apiKeyKey;

  @Bean
  @ConditionalOnProperty(name = "openapipetstore.security.apiKey.key")
  public ApiKeyRequestInterceptor apiKeyRequestInterceptor() {
    return new ApiKeyRequestInterceptor("header", "api_key", this.apiKeyKey);
  }

  public static class OAuth2RequestInterceptor implements RequestInterceptor {

    private final OAuth2AuthorizedClientManager oAuth2AuthorizedClientManager;
    private final OAuth2AuthorizeRequest oAuth2AuthorizeRequest;

    public OAuth2RequestInterceptor(OAuth2AuthorizeRequest oAuth2AuthorizeRequest,OAuth2AuthorizedClientManager oAuth2AuthorizedClientManager){
      this.oAuth2AuthorizeRequest = oAuth2AuthorizeRequest;
      this.oAuth2AuthorizedClientManager = oAuth2AuthorizedClientManager;
    }

    @Override
    public void apply( final RequestTemplate template ) {
      template.header( HttpHeaders.AUTHORIZATION, getBearerToken() );
    }

    public OAuth2AccessToken getAccessToken() {
      final OAuth2AuthorizedClient authorizedClient = oAuth2AuthorizedClientManager.authorize(oAuth2AuthorizeRequest);
      if (authorizedClient == null) {
        throw new OAuth2AuthenticationException( "Client failed to authenticate");
      }
     return authorizedClient.getAccessToken();
    }

    public String getBearerToken() {
      final OAuth2AccessToken accessToken = getAccessToken();
      return String.format( "%s %s", accessToken.getTokenType().getValue(), accessToken.getTokenValue() );
    }
  }

}

File: samples/server/petstore/kotlin-spring-cloud-4/src/main/kotlin/org/openapitools/configuration/ClientConfiguration.kt
package org.openapitools.configuration

import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty



import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

import org.springframework.security.authentication.AnonymousAuthenticationToken
import org.springframework.security.oauth2.client.AuthorizedClientServiceOAuth2AuthorizedClientManager
import org.springframework.security.oauth2.client.OAuth2AuthorizeRequest
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientManager
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService
import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository
import org.springframework.security.oauth2.core.OAuth2AuthenticationException
import org.springframework.security.oauth2.core.OAuth2AccessToken
import org.springframework.security.core.authority.AuthorityUtils
import org.springframework.http.HttpHeaders

import feign.RequestInterceptor
import feign.RequestTemplate

@Configuration
class ClientConfiguration {


    @Bean
    @ConditionalOnProperty(prefix = "spring.security.oauth2.client.registration.petstore_authImplicit", name = ["enabled"], havingValue = "true")
    fun implicitOAuth2RequestInterceptor(implicitAuthorizedClientManager: OAuth2AuthorizedClientManager): OAuth2RequestInterceptor {
        return OAuth2RequestInterceptor(
            OAuth2AuthorizeRequest.withClientRegistrationId("petstore_authImplicit")
                .principal(AnonymousAuthenticationToken(CLIENT_PRINCIPAL_IMPLICIT, CLIENT_PRINCIPAL_IMPLICIT, AuthorityUtils.createAuthorityList("ROLE_ANONYMOUS")))
                .build(),
            implicitAuthorizedClientManager
        )
    }

    @Bean
    @ConditionalOnProperty(prefix = "spring.security.oauth2.client.registration.petstore_authImplicit", name = ["enabled"], havingValue = "true")
    fun implicitAuthorizedClientManager(
        clientRegistrationRepository: ClientRegistrationRepository,
        authorizedClientService: OAuth2AuthorizedClientService
    ): OAuth2AuthorizedClientManager {
        return AuthorizedClientServiceOAuth2AuthorizedClientManager(clientRegistrationRepository, authorizedClientService)
    }

    @Value("\${openapipetstore.security.api_key.key:}")
    private lateinit var apiKeyKey: String

    @Bean
    @ConditionalOnProperty("openapipetstore.security.api_key.key")
    fun apiKeyRequestInterceptor(): ApiKeyRequestInterceptor {
        return ApiKeyRequestInterceptor("header", "api_key", this.apiKeyKey)
    }

    class OAuth2RequestInterceptor(
        private val oAuth2AuthorizeRequest: OAuth2AuthorizeRequest,
        private val oAuth2AuthorizedClientManager: OAuth2AuthorizedClientManager
    ) : RequestInterceptor {

        override fun apply(template: RequestTemplate) {
            template.header(HttpHeaders.AUTHORIZATION, getBearerToken())
        }

        fun getAccessToken(): OAuth2AccessToken {
            val authorizedClient = oAuth2AuthorizedClientManager.authorize(oAuth2AuthorizeRequest)
                ?: throw OAuth2AuthenticationException("Client failed to authenticate")
            return authorizedClient.accessToken
        }

        fun getBearerToken(): String {
            val accessToken = getAccessToken()
            return String.format(java.util.Locale.ROOT, "%s %s", accessToken.tokenType?.value, accessToken.tokenValue)
        }
    }

    companion object {
        private const val CLIENT_PRINCIPAL_IMPLICIT = "oauth2FeignClient"
    }
}
